### ISSUE TITLE
String Indexer Assignable to Generic Mapped Type

### ISSUE THREAD

**Issue by jack-williams** [Microsoft/TypeScript#28798]
**TypeScript Version:** Version 3.3.0-dev

**Search Terms:** string indexer generic mapped type assignable

**Code**
This seems like a regression (if it is a bug). I get the expected behaviour on the playground 3.1.

```ts
function tester<K extends 'a'>(x: Record<K, number>, y: { [x: string]: number }) {
    x = y; // no error: not expected
    y = x; // no error: expected
}

function tester2(x: Record<'a', number>, y: { [x: string]: number }) {
    x = y; // error: expected
    y = x; // no error: expected
}
```

**Expected behavior:**

A string indexer should not be assignable to a mapped type with a generic constraint.

**Actual behavior:**

String indexer appears to be assignable to generic record type.

**Playground Link:** [playground with expected behaviour](https://www.typescriptlang.org/play/#src=function%20tester%3CK%20extends%20'a'%3E(x%3A%20Record%3CK%2C%20number%3E%2C%20y%3A%20%7B%20%5Bx%3A%20string%5D%3A%20number%20%7D)%20%7B%0D%0A%20%20%20%20x%20%3D%20y%3B%20%2F%2F%20error%3A%20expected%0D%0A%20%20%20%20y%20%3D%20x%3B%20%2F%2F%20no%20error%3A%20expected%0D%0A%7D)

**Update**. I think this was introduced by #28218.

```ts
// A source type T is related to a target type { [P in Q]: X } if Q is related to keyof T and T[Q] is related to X.
if (!isGenericMappedType(source) && isRelatedTo(getConstraintTypeFromMappedType(target), getIndexType(source))) {
    const indexedAccessType = getIndexedAccessType(source, getTypeParameterFromMappedType(target));
    const templateType = getTemplateTypeFromMappedType(target);
    if (result = isRelatedTo(indexedAccessType, templateType, reportErrors)) {
        return result;
    }
}
```

In this instance we have `'a'` is related `string`, but a type with a key set of type `string` is not guaranteed to have the key `a`. When relating keys I think `string` and `number` should really be bottom types, so a possible fix is to filter them from the source index type:

```ts
filterType(getIndexType(source), t => !(t.flags & (TypeFlags.String | TypeFlags.Number)
// and symbol??
```

**Comment by ahejlsberg** [comment]
@jack-williams I think that is a very sensible suggestion.

**Comment by jack-williams** [comment]
If this is low down on the prio list I'd be happy to give fixing this a go.

**Comment by jack-williams** [comment]
Fixed here: #30769
